cmd/crc/cmd/start.go

```diff
@@ -155,3 +157,21 @@ func checkIfNewVersionAvailable(noUpdateCheck bool) {
 	}
 	logging.Debugf("No new version available. The latest version is %s", newVersion)
 }
+
+func handleProxyConfig() (*network.ProxyConfig, error) {
+	httpProxy := crcConfig.GetString(config.HttpProxy.Name)
+	httpsProxy := crcConfig.GetString(config.HttpsProxy.Name)
+	noProxy := crcConfig.GetString(config.NoProxy.Name)
+
+	proxyConfig, err := network.NewProxyConfig(httpProxy, httpsProxy, noProxy)
+
```

Comment:
have you considered having `ProxyConfig` do the `crcConfig.GetString` calls itself rather than doing it here? We want *all* users of ProxyConfig to get values from crc config if they are available, so better to do it from the `ProxyConfig` code?

Reply:
@cfergeau I didn't get you? you mean move this to package level and have something following? if this is what have in mind then no since `config` is coming from `cmd` side not from the `pkg` side so I don't want to have anything which come from `cmd` to `pkg` otherwise moving forward we might hit the cyclic dependencies. 

```golang
func GetProxyConfig() (*ProxyConfig, error) {
	defaultScheme := "http"
        
        httpProxy := crcConfig.GetString(config.HttpProxy.Name)
	if httpProxy == "" {
		httpProxy = getProxyFromEnv("http_proxy")
	}
	err := ValidateProxyURL(httpProxy, defaultScheme)
	if err != nil {
		return nil, err
	}
```

Reply:
also not following the argument, but as pointed out we avoid moving code into pkg that handles with basic config setup.

Reply:
`config` is coming from `pkg`: `import crcConfig "github.com/code-ready/crc/pkg/crc/config"`
Requiring every user of the ProxyConfig package to know that at creation time, you need to pass it configuration values you fetched yourself is awkward/error-prone.

Reply:
Alternative which might work: add `DefaultHttpProxy`, `DefaultHttpsProxy`, `DefaultNoProxy` variables to the proxy package, use these in `NewProxyConfig` instead of its 3 args, and set `Default*` from the config in the start command.

Reply:
@cfergeau I am still missing your point of view but I am talking about `config.HttpProxy.Name` part where we get the value of it. This `config` is part of `cmd` and not `pkg` or you are suggesting something like?
```
httpProxy := crcConfig.GetString("http-proxy") => using directly the 'http-proxy' string instead of getting it from the configs?
```

Reply:
> Alternative which might work: add DefaultHttpProxy, DefaultHttpsProxy, DefaultNoProxy variables to the proxy package, use these in NewProxyConfig instead of its 3 args, and set Default* from the config in the start command.

@cfergeau you mean like below?  how does that make difference, I still not sure what is missing? 

```golang
func handleProxyConfig() (*network.ProxyConfig, error) {
	proxyConfig.DefaultHttpProxy := crcConfig.GetString(config.HttpProxy.Name)
	[...]
	proxyConfig, err := network.NewProxyConfig()
```

Reply:
>  but I am talking about config.HttpProxy.Name part where we get the value of it.
Ah, right, I missed this one :) I would define it in dns/proxy.go, this way any user of the proxy API has easy access to the config keys to use. But apparently we only want `cmd/` to make use of `pkg/crc/config`?

Reply:
> > Alternative which might work: add DefaultHttpProxy, DefaultHttpsProxy, DefaultNoProxy variables to the proxy package, use these in NewProxyConfig instead of its 3 args, and set Default* from the config in the start command.
> 
> @cfergeau you mean like below? how does that make difference, I still not sure what is missing?
> 
> ```go
> func handleProxyConfig() (*network.ProxyConfig, error) {
> 	proxyConfig.DefaultHttpProxy := crcConfig.GetString(config.HttpProxy.Name)
> 	[...]
> 	proxyConfig, err := network.NewProxyConfig()
> ```

In `init()` or some function called early, we'd initialize proxy from config, either by directly setting the `Default*Proxy` variables, or through a dedicated helper (`proxyConfig.InitProxyDefaults()`).
Once you've done that, the `handleProxyConfig()` delegate is no longer needed, any code can call `proxyConfig, err := network.NewProxyConfig()` directly and check for errors. The `Default*Proxy` variables are just there if we want to avoid reading the config from `pkg/`

Reply:
> But apparently we only want cmd/ to make use of pkg/crc/config?

@cfergeau No, that is not the case but there is also a `config` which is part of `cmd/config` which we don't want to use in `pkg` level.

>Once you've done that, the handleProxyConfig() delegate is no longer needed, any code can call proxyConfig, err := network.NewProxyConfig() directly and check for errors.

Doesn't this will make everything implicit? Right now we know only start and setup need the proxy and no other cmd really required it. Now from the start side if any other function require then we can pass this as part of function parameter (which will make it explicit) instead directly invoking the `network.NewProxyConfig` from that function. I think it just matter of taste I would like to have it passed explicitly.

Reply:
> Doesn't this will make everything implicit?

It makes sense to me that a proxy is used whenever one is needed rather than having to be explicitly provided before proxy can be used. For example, cavaliercoder/grab uses http_proxy by default.

>  I think it just matter of taste I would like to have it passed explicitly.

My main objection is that this is currently done through a function delegate, as long as it's replaced with something simpler, I'm fine with both explicit and implicit.

Reply:
Since it's unused at the moment, I'd just remove it from that commit so that we move forward with it, and rediscuss that when there is an actual usecase for it.

Reply:
> Since it's unused at the moment, I'd just remove it from that commit so that we move forward with it, and rediscuss that when there is an actual usecase for it.

@cfergeau it is used for setup command atm.